---
title: "clark_custom_visual"
output: html_document
---

# set the working directory to the Dropbox path where data files are storing at
```{r setup, include=FALSE}
custom_path="/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/"

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = custom_path)
```
###include a color palette for ggplot
```{r}
manual_color <-c("#F94144","#80B918","#FEE440","#4361EE","#FFADAD","#FFD6A5","#FDFFB6","#CAFFBF","#9BF6FF","#A0C4FF","#BDB2FF","#FFC6FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C")

```

# Libraries imported for analysis
```{r}

library(dplyr)
library(magrittr)
library(tidyverse)
library(plotly)
library(rstatix)


```

### since only species level classification has been done for clark (with rat analysis), we will extract Domain information with clark's species results

```{r}

species_path <- paste(custom_path, "species_custom/" , sep="")
files <- list.files(species_path)



i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(species_path, file,sep=""), header = TRUE)
  # get the domain column
  taxa_count <- sample_record %>% group_by(Lineage) %>% mutate(Domain = sum(Count)) %>% select("Lineage","Domain")
  taxa_count <- unique(taxa_count)
  colnames(taxa_count) <- c("Name","Count")
  
  # taxa_count %<>% subset(Name != "Chordata" & Name != "UNKNOWN")
  rownames(taxa_count) <- NULL # reset row number
  colnames(taxa_count) <- c("Name", sample_name)
  if (i == 0){
    domain_full_table <- taxa_count
  } else {
    domain_full_table <- full_join(domain_full_table, taxa_count,by = "Name")
  }
  i = i + 1
}

domain_full_table[is.na(domain_full_table)] <- 0
domain_full_table
 
sample_keys <- colnames(domain_full_table)[-1]
domain_full_table %<>% gather(sample_keys,key="Samples", value="Read_Count")

order <- unique(domain_full_table$Samples)
domain_full_table$Samples <- factor(domain_full_table$Samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))

domain_full_table$Name <- factor(domain_full_table$Name, levels = c("UNKNOWN", "Eukaryota","Bacteria","Viruses","Archaea"))

domain_full_table %>% group_by(Samples) %>% 
  mutate(Total = sum(Read_Count), Percentage = Read_Count/Total) -> domain_major

# Total number of reads for each sample (unclassified + three domains + other sequences)
domain_full_table %>% group_by(Samples) %>% summarise(total_reads=sum(Read_Count)) -> clark_total_reads
as.data.frame(list(mean_reads=mean(clark_total_reads$total_reads),std=sd(clark_total_reads$total_reads))) -> clark_total_reads_summary # avg number of reads across sample
clark_total_reads_summary

```

# Plot Domain level taxaonomy count with UNKNOWN and Eukaryota to have an overall picture of the samples
```{r}
plot_domain <- ggplot(domain_major, aes(x=Samples, y=Read_Count, fill=Name)) +
  geom_bar(stat="identity") + 
  theme(axis.text.x=element_text(angle=90)) +
  labs(title = "Metagenomic Reads Classification with Clark DB (Domain)", y= "Number of Reads",
       x="Samples")+
  scale_fill_manual(values=manual_color)

plot_domain
ggplotly(plot_domain)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Absolute/clark_cus_domain.png",width = 8, height = 6)
########Relative Plot#################

domain_relative <- ggplot(domain_major, aes(x=Samples, y=Percentage, fill=Name)) +
  geom_bar(stat="identity") + 
  theme(axis.text.x=element_text(angle=90)) +
  labs(title = "Metagenomic Relative Abundance Classification with Clark DB (Domain)",
       x="Samples") +  scale_y_continuous(labels = scales::percent_format()) +
  scale_fill_manual(values = manual_color)
ggplotly(domain_relative)
domain_relative
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Relative/clark_cus_domain_rel.png",width = 8, height = 6)

```


# Plot Phylum level abundance without Eukaryota and Unknown information for better visualization 
```{r}
species_path <- paste(custom_path, "species_custom/" , sep="")
files <- list.files(species_path)


i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(species_path, file,sep=""), header = TRUE)
  # get the domain column
  taxa_count <- sample_record %>% group_by(X) %>% mutate(Phylum = sum(Count)) %>% select("X","Phylum") #%>% subset(Phylum >= 100)
  taxa_count <- unique(taxa_count)
  colnames(taxa_count) <- c("Name","Count")
  
  taxa_count %<>% subset(Name != "Chordata" & Name != "")
  rownames(taxa_count) <- NULL # reset row number
  colnames(taxa_count) <- c("Name", sample_name)
  if (i == 0){
    phylum_full_table <- taxa_count
  } else {
    phylum_full_table <- full_join(phylum_full_table, taxa_count,by = "Name")
  }
  i = i + 1
}

phylum_full_table[is.na(phylum_full_table)] <- 0
phylum_full_table
 
sample_keys <- colnames(phylum_full_table)[-1]
phylum_full_table %<>% gather(sample_keys,key="Samples", value="Read_Count")

all_phylums <- unique(phylum_full_table$Name) # all phylums classified
# phylums I want to factored into order so they can be assigned with a specific color during plotting


order_phylum <- c("Proteobacteria","Actinobacteria","Cyanobacteria","Firmicutes","Bacteroidetes","Uroviricota","Tenericutes","Spirochaetes","Aquificae","Artverviricota","Peploviricota","Planctomycetes") 

remaining_phylum <- all_phylums %>% data.frame() %>% subset(!c(all_phylums %in% order_phylum))
remaining_phylum <- remaining_phylum$.

phylum_full_table$Samples <- factor(phylum_full_table$Samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
phylum_full_table$Name <- factor(phylum_full_table$Name, level=c(order_phylum,remaining_phylum))

phylum_full_table %>% group_by(Samples) %>% 
  mutate(Total = sum(Read_Count), Percentage = Read_Count/Total) %>%
  subset(Percentage >= 0.001)-> phylum_major

phylum_major


```



# Plot Domain level taxaonomy count with UNKNOWN and Eukaryota to have an overall picture of the samples
```{r}

# color palette adjusted to match with Kraken2 color code for genus taxa
manual_color <-c("#F94144","#80B918","#FEE440","#4361EE","#FFADAD","#FFD6A5","#FDFFB6","#CAFFBF","#BDB2FF","#FCCA46","#A1C181","#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C")

phylum_plot <- ggplot(phylum_major, aes(x=Samples,y=Read_Count, fill=Name)) + 
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=90),legend.text = element_text(size = 8),legend.key.size = unit(0.4,"cm"))+
  labs(title = "Clark Microbiome Reads Profile (Phylum)", y= "Number of Reads", x="Samples" )  +
  scale_fill_manual(values=manual_color)

phylum_plot
# ggplotly(phylum_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Absolute/clark_cus_phylum.png", width = 8, height = 6)

# ggplotly(plot_phylum)

########Relative Plot#################


phylum_relative <- ggplot(phylum_major, aes(x=Samples,y=Percentage, fill=Name)) + 
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=90),legend.text = element_text(size = 8),legend.key.size = unit(0.4,"cm"))+
  labs(title = "Clark Microbiome Composition Profile (Phylum)", y= "Percentage", x="Samples")+scale_y_continuous(labels = scales::percent_format()) +
  scale_fill_manual(values = manual_color)

phylum_relative
# ggplotly(phylum_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Relative/clark_cus_phylum_rel.png", width = 8, height = 6)


```
### genus level classification 
```{r}
species_path <- paste(custom_path, "species_custom/" , sep="")
files <- list.files(species_path)


i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(species_path, file,sep=""), header = TRUE)
  # get the domain column
  taxa_count <- sample_record %>% group_by(X.4) %>% mutate(Genus = sum(Count)) %>% select("X.4","Genus") %>% subset(Genus >= 300)
  taxa_count <- unique(taxa_count)
  colnames(taxa_count) <- c("Name","Count")
  
  taxa_count %<>% subset(Name != "Homo" & Name != "Rattus" & Name != "")
  rownames(taxa_count) <- NULL # reset row number
  colnames(taxa_count) <- c("Name", sample_name)
  if (i == 0){
    genus_full_table <- taxa_count
  } else {
    genus_full_table <- full_join(genus_full_table, taxa_count,by = "Name")
  }
  i = i + 1
}

genus_full_table[is.na(genus_full_table)] <- 0
# genus_full_table
 
sample_keys <- colnames(genus_full_table)[-1]
genus_full_table %<>% gather(sample_keys,key="Samples", value="Read_Count")





# genus I want to factored in order so they can be assigned with a specific color during plotting
all_genus <- unique(genus_full_table$Name) # allgenus classified
order_genus <- c("Streptomyces","Andhravirus","Mycoplasma","Spiroplasma","Lactobacillus","Bordetella","Leptospira","Yersinia","Escherichia","Bartonella","Bacteroides","Microcystis","Nostoc","Alistipes","Anabaena","Faecalibacterium","Prevotella") 
remaining_genus <- all_genus %>% data.frame() %>% subset(!c(all_genus %in% order_genus)) # deselect genus subsetted to put in order
remaining_genus <- remaining_genus$. # data.frame to vector

order <- unique(genus_full_table$Samples)
genus_full_table$Samples <- factor(genus_full_table$Samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
genus_full_table$Name <- factor(genus_full_table$Name, level=c(order_genus,remaining_genus))

genus_full_table%>% group_by(Samples) %>% 
  mutate(Total = sum(Read_Count), Percentage = Read_Count/Total) %>%
  subset(Percentage >= 0.01)->major_genus
```


### Plot Genus level absolute abundance
```{r}
# color palette adjusted to match with Kraken2 color code for genus taxa

manual_color <-c("#4361EE","#FFD6A5","#FDFFB6","#CAFFBF","#BDB2FF","#FE7F2D","#FCCA46","#A1C181","#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#50514F","#FFE066","#70C1B3","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff","#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C")

genus_plot <- ggplot(major_genus, aes(x=Samples,y=Read_Count, fill=Name)) + 
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=90),legend.key.size = unit(0.4,"cm"), legend.text = element_text(size = 8),plot.caption = element_text(hjust = 0, face= "italic")) +
  labs(title = "Metagenomic Reads Classification with Clar(with Rat) DB (Genus)", y= "Number of Reads",
       x="Samples", caption= "*For the purpose of visualization, only taxa classified with more than 10% in composition will be shown") +
  scale_fill_manual(values=manual_color)

genus_plot

# ggplotly(genus_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Absolute/clark_cus_genus.png")

########Relative Plot#################


genus_relative <- ggplot(major_genus, aes(x=Samples,y=Percentage, fill=Name)) + 
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=90),legend.text = element_text(size = 8),legend.key.size = unit(0.4,"cm"),plot.caption = element_text(hjust = 0, face= "italic"))+
  labs(title = "Metagenomic Relative Abundancen with Clark(with Rat) DB (Genus)", y= "Percentage", x="Samples" , caption= "*For the purpose of visualization, only taxa classified with more than 10% in composition will be shown")+scale_y_continuous(labels = scales::percent_format()) +
  scale_fill_manual(values = manual_color)

genus_relative
# ggplotly(phylum_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Relative/clark_cus_genus_rel.png")


```
