---
title: "custom_kraken2_output_visual"
author: "Rachel Xu"
date: "6/22/2020"
output: html_document
---
### PATH to custom output result:
```{r}
path = "/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/"
setwd(path)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = path)
```

```{r}
library(dplyr)
library(magrittr)
library(ggplot2)
library(tidyverse)
library(plotly)
library("phyloseq")
```
###include a color palette for ggplot
```{r}
manual_color <-c("#F94144","#80B918","#FEE440","#4361EE","#FFADAD","#FFD6A5","#FDFFB6","#CAFFBF","#9BF6FF","#A0C4FF","#BDB2FF","#FFC6FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C")

```
## Kraken2 kreport:
    **1) Absolute Reads**
    **2) root: celluar organism (Eukaryota + Bacteria) + Viruses**
    **3) all_reads: Eukaryota + Bacteria + Viruses + unclassified**
    
### Domain Classification 
```{r}
absolute_path = paste(path,"/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  # add "| trimws(V6) == "other sequences"" to subset if counting total number of reads
  domain_absolute <- current_sample %>% subset(V4 == "D" | V4 == "U" | trimws(V6) == "other sequences") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))

  colnames(domain_absolute) <- c("Domain", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    domain_all_samples <-  domain_absolute

  } else {
    domain_all_samples <- full_join(domain_all_samples, domain_absolute,by="Domain")

  }
  i = i + 1
}

# save(domain_all_samples, file ="custom_absolute_domain.Rdata")

domain_all_samples[is.na(domain_all_samples)] <- 0
keys_to_gather <- colnames(domain_all_samples)[-1] # gather samples, exclude "Domain" from colnames

# prepare dataframe for ggplot
domain_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")


domain_all_samples$samples <- factor(domain_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
domain_all_samples$Domain <- factor(domain_all_samples$Domain, level=unique(domain_all_samples$Domain))

# Total number of reads for each sample (unclassified + three domains + other sequences)
domain_all_samples %>% group_by(samples) %>% summarise(total_reads=sum(Reads)) -> summary_domain
c(mean(summary_domain$total_reads),sd(summary_domain$total_reads)) # avg number of reads across sample

# other sequences summary
domain_all_samples %>% subset(Domain == "other sequences") %>% summarise(mean_other=mean(Reads),std=sd(Reads)) 
```

```{r}
plot <- ggplot(domain_all_samples, aes(x=samples,y=Reads,fill=Domain)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Reads Classification with Customized Kraken2 DB (Domain)", y="Number of Reads", x = "Samples")+
  theme(axis.text.x=element_text(angle=90)) +
  scale_fill_manual(values=manual_color)

plot
ggplotly(plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Absolute/kraken2_cus_domain.png", width = 8, height=6)

#### plot relative abundance 
domain_all_samples %>%
  group_by(samples) %>%
  mutate(Total=sum(Reads)) %>%
  mutate(Relative=Reads/Total) -> domain_relative


relative_plot <- ggplot(domain_relative, aes(x=samples,y=Relative,fill=Domain)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Relative Abundance with Customized Kraken2 DB (Domain)", y="Percentage", x = "Samples")+
  scale_y_continuous(labels = scales::percent_format()) +
  theme(axis.text.x=element_text(angle=90))+
  scale_fill_manual(values=manual_color)
ggplotly(relative_plot)
relative_plot
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Relative/kraken2_cus_domain_rel.png", width = 8, height=6)

domain_relative
domain_relative %>% group_by(Domain) %>% summarise(mean_reads = mean(Reads), std_reads = sd(Reads), mean_Percentage= mean(Relative), std_Percentage=sd(Relative))
```

### Classify phylum, excluding unclassified and Eukaryote reads to increase local resolution
    
```{r}
absolute_path = paste(path,"custom/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  phylum_absolute <- current_sample %>% subset(V4 == "P") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))
  phylum_absolute <- phylum_absolute %>% subset(c(trimed!="Chordata")) #%>% mutate( V2 >= 100) # deselect phylum belongs to Eukrayota
  colnames(phylum_absolute) <- c("Phylum", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    phylum_all_samples <-  phylum_absolute

  } else {
    phylum_all_samples <- full_join(phylum_all_samples, phylum_absolute,by="Phylum")

  }
  i = i + 1
}

# save(phylum_all_samples, file ="standard_absolute_phylum_micro_only.Rdata")

phylum_all_samples[is.na(phylum_all_samples)] <- 0
keys_to_gather <- colnames(phylum_all_samples)[-1] # gather samples, exclude "Phylum" from colnames

# prepare dataframe for ggplot
phylum_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

all_phylums <- unique(phylum_all_samples$Phylum) # all phylums classified
# phylums I want to factored into order so they can be assigned with a specific color during plotting
order_phylum <- c("Proteobacteria","Actinobacteria","Cyanobacteria","Firmicutes","Bacteroidetes","Uroviricota","Tenericutes","Spirochaetes","Fusobacteria","Chlamydiae","Aquificae","Chloroflexi","Thermotogae","Artverviricota","Peploviricota","Deinococcus-Thermus","Planctomycetes") 
remaining_phylum <- all_phylums %>% data.frame() %>% subset(!c(all_phylums %in% order_phylum)) 
remaining_phylum <- remaining_phylum$.

phylum_all_samples$samples <- factor(phylum_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
phylum_all_samples$Phylum <- factor(phylum_all_samples$Phylum, level=c(order_phylum,remaining_phylum))


# subsetting taxonomy with more than 1% in relative composition from each sample
phylum_all_samples %>% group_by(samples) %>%
  mutate(Total=sum(Reads), Percentage = Reads/Total) %>% subset(Percentage >= 0.001) ->major_phylum



```

```{r}


plot <- ggplot(major_phylum, aes(x=samples,y=Reads,fill=Phylum)) +
  geom_bar(stat="identity") + 
  labs(title = "Kraken2 Microbiome Reads Profile with Customized DB (Phylum)", y="Number of Reads", x = "Samples") + 
   theme(axis.text.x = element_text(angle=90),legend.text = element_text(size = 8),legend.key.size = unit(0.4,"cm"))+
  scale_fill_manual(values=manual_color)

plot

ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Absolute/kraken2_cus_phylum.png",width=8, height=6)




relative_plot <- ggplot(major_phylum, aes(x=samples,y=Percentage,fill=Phylum)) +
  geom_bar(stat="identity") + 
  labs(title = "Kraken2 Microbiome Composition Profile with Customized DB (Phylum)", y="Percentage", x = "Samples") + 
 theme(axis.text.x = element_text(angle=90),legend.text = element_text(size = 8),legend.key.size = unit(0.4,"cm")) +
  scale_y_continuous(labels = scales::percent_format()) +
  theme(axis.text.x=element_text(angle=90)) + 
  scale_fill_manual(values=manual_color)
relative_plot
# ggplotly(relative_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Relative/kraken2_cus_phylum_rel.png",width=8, height=6)

```

### Classify genus, excluding unclassified and Eukaryote reads to increase local resolution

```{r}
absolute_path = paste(path,"custom/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  genus_absolute <- current_sample %>% subset(V4 == "G") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))
  genus_absolute <- genus_absolute #%>% subset(trimed!="Rattus" & trimed !="Homo") #%>% subset(V2 >= 300) # deselect genus belongs to Eukrayota
  colnames(genus_absolute) <- c("genus", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    genus_all_samples <-  genus_absolute

  } else {
    genus_all_samples <- full_join(genus_all_samples, genus_absolute,by="genus")

  }
  i = i + 1
}

# save(genus_all_samples, file ="standard_absolute_genus_micro_only.Rdata")

genus_all_samples[is.na(genus_all_samples)] <- 0
keys_to_gather <- colnames(genus_all_samples)[-1] # gather samples, exclude "genus" from colnames

# prepare dataframe for ggplot
genus_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

# genus I want to factored in order so they can be assigned with a specific color during plotting
all_genus <- unique(genus_all_samples$genus) # allgenus classified
order_genus <- c("Homo","Rattus","Xanthomonas","Pseudomonas","Pasteurella","Streptomyces","Calothrix","Andhravirus","Mycoplasma","Spiroplasma","Bacillus","Staphylococcus","Lactobacillus","Clostridium","Mycobacterium","Bordetella","Leptospira","Yersinia","Escherichia","Bartonella","Bacteroides","Microcystis","Nostoc","Mycobacteroides","Alistipes","Parabacteroides","Anabaena","Dolichospermum","Faecalibacterium","Sphaerospermopsis","Prevotella","Achromobacter") 
remaining_genus <- all_genus %>% data.frame() %>% subset(!c(all_genus %in% order_genus)) # deselect genus subsetted to put in order
remaining_genus <- remaining_genus$. # data.frame to vector

genus_all_samples$samples <- factor(genus_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
genus_all_samples$genus <- factor(genus_all_samples$genus, level=c(order_genus,remaining_genus))

# subsetting taxonomy with more than 1% in relative composition from each sample 
genus_all_samples %>% group_by(samples) %>%
  mutate(Total=sum(Reads), Percentage = Reads/Total) %>% subset(Percentage >= 0.010) ->major_genus

# number of reads classified into microbiome composition (exclude homo and rattus)
genus_all_samples %>% group_by(samples) %>%
  mutate(Total=sum(Reads), Percentage = Reads/Total) %>%subset(genus != "Rattus" & genus !="Homo") %>% summarise(sum_reads=sum(Reads), sum_percentage=sum(Percentage)) -> cus_micro_comp
cus_micro_comp
```

```{r}
manual_color <-c("#54478C","#97CBF0ff","#F94144","#80B918","#FEE440","#4361EE","#FFADAD","#FFD6A5","#FDFFB6","#CAFFBF","#9BF6FF","#A0C4FF","#BDB2FF","#FFC6FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff")

plot <- ggplot(major_genus, aes(x=samples,y=Reads,fill=genus)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Reads Classification with Customized Kraken2 DB (Genus)", y="Number of Reads", x = "Samples") + 
  theme(legend.text = element_text(size =8 ),legend.key.size =unit(0.4,"cm"))+
  scale_fill_manual(values=manual_color)
ggplotly(plot)
plot
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Absolute/kraken2_cus_genus.png",width=10, height=6)



relative_plot <- ggplot(major_genus, aes(x=samples,y=Percentage,fill=genus)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Relative Abundance with Customized Kraken2 DB (Genus)", y="Percentage", x = "Samples") + 
  theme(legend.text = element_text(size =8 ),legend.key.size =unit(0.4,"cm")) +
  scale_y_continuous(labels = scales::percent_format()) +
  theme(axis.text.x=element_text(angle=90)) + 
  scale_fill_manual(values=manual_color)
relative_plot
ggplotly(relative_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Relative/kraken2_cus_genus_rel.png",width=10, height=6)


```