---
title: "minikraken_analysis"
author: "Rachel Xu"
date: "6/11/2020"
output: html_document
---
### PATH to minikraken output result:
```{r}
path = "/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/minikraken/"
setwd(path)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = path)
```

```{r}
library(dplyr)
library(magrittr)
library(ggplot2)
library(tidyverse)
library(plotly)
```
###include a color palette for ggplot
```{r}
manual_color <-c("#4361EE","#9BF6FF","#FFADAD","#FDFFB6","#FFD6A5","#CAFFBF","#A0C4FF","#BDB2FF","#FFC6FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C","#F94144","#80B918","#FEE440")

```

Kraken2 kreport:
    1) Absolute Reads
    2) root: celluar organism (Eukaryota + Bacteria) + Viruses
    3) all_reads: Eukaryota + Bacteria + Viruses + unclassified 
```{r}
mini_path <- paste(path,"absolute/",sep="")
files <- list.files(mini_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(mini_path,file,sep=""),sep="\t",header=FALSE)
  # add "| trimws(V6) == "other sequences"" to subset if counting total number of reads
  domain_absolute <- current_sample %>% subset(V4 == "D" | V4 == "U" | trimws(V6) == "other sequences") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))
  
  colnames(domain_absolute) <- c("Domain", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    domain_all_samples <-  domain_absolute
    
  } else {
    domain_all_samples <- full_join(domain_all_samples, domain_absolute,by="Domain")
    
  }
  i = i + 1
}

save(domain_all_samples, file ="minikraken_absolute_domain.Rdata")

domain_all_samples[is.na(domain_all_samples)] <- 0
keys_to_gather <- colnames(domain_all_samples)[-1] # gather samples, exclude "Domain" from colnames

# prepare dataframe for ggplot
domain_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

order <- unique(domain_all_samples$samples)
domain_all_samples$samples <- factor(domain_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
domain_all_samples$Domain <- factor(domain_all_samples$Domain, level=c("unclassified", "other sequences","Eukaryota","Bacteria","Viruses","Archaea"))


# Total number of reads for each sample (unclassified + three domains + other sequences)
domain_all_samples %>% group_by(samples) %>% summarise(total_reads=sum(Reads)) -> mini_total_reads
as.data.frame(list(mean_reads= mean(mini_total_reads$total_reads),std=sd(mini_total_reads$total_reads))) -> mini_total_reads_summary# avg number of reads across sample

# other sequences summary
domain_all_samples %>% subset(Domain == "other sequences") -> mini_other_sequences
mini_other_sequences %>%summarise(mean_other=mean(Reads),std=sd(Reads)) -> mini_other_sequences_summary


mini_total_reads_summary# avg number of reads across sample
mini_other_sequences_summary # avg number of reads classified into "other sequences"

```

```{r}

plot <- ggplot(domain_all_samples, aes(x=samples,y=Reads,fill=Domain)) +
  geom_bar(stat="identity") + 
 theme_classic() +
  theme(legend.key.size = unit(0.6, "cm"), legend.text = element_text(size=12),legend.title = element_text(size=12,face="bold"), axis.text.y = element_text(face="bold", size=10),axis.title = element_blank(), axis.text.x= element_blank(),legend.position="none") +
  scale_fill_manual(values=manual_color)

ggplotly(plot)
plot
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Absolute/kraken2_mini_domain.png",width=8,height=6)

#### plot relative abundance 
domain_all_samples %>%
  group_by(samples) %>%
  mutate(Total=sum(Reads)) %>%
  mutate(Relative=Reads/Total) -> domain_relative

relative_plot <- ggplot(domain_relative, aes(x=samples,y=Relative,fill=Domain)) +
  geom_bar(stat="identity") + 
  theme_classic() +
  theme(legend.key.size = unit(0.6, "cm"), legend.text = element_text(size=12),legend.title = element_text(size=12,face="bold"), axis.text.y = element_text(face="bold", size=10),axis.title = element_blank(), axis.text.x= element_blank(),legend.position="none") +
  scale_fill_manual(values=manual_color)+
  scale_y_continuous(labels = scales::percent_format())

ggplotly(relative_plot)
relative_plot
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Relative/kraken2_mini_domain_rel.png",width=8,height=6)

# classified profile summary 
domain_relative
domain_relative %>% group_by(Domain) %>% summarise(mean_Reads=mean(Reads), std_Reads=sd(Reads), mean_percentage=mean(Relative), std_percentage=sd(Relative))
```

### Compare to classification results with standard database, the amount of reads classified with minikraken database was much lesser. Thus minikraken2 is not able to produce a reliable results compare to using kraken2's standard databases