---
title: "standard_kraken2_output_visual"
author: "Rachel Xu"
date: "6/15/2020"
output: html_document
---
### PATH to minikraken output result:
```{r}
path = "/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/standard"
setwd(path)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = path)
```

```{r}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("phyloseq")
library(dplyr)
library(magrittr)
library(ggplot2)
library(tidyverse)
library(plotly)
library(phyloseq)
```

###Color Palette for plotting
```{r}

manual_color <-c("#F94144","#80B918","#FEE440","#4361EE","#FFADAD","#FFD6A5","#FDFFB6","#CAFFBF","#9BF6FF","#A0C4FF","#BDB2FF","#FFC6FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C")
```
## Kraken2 kreport:
    **1) Absolute Reads**
    **2) root: celluar organism (Eukaryota + Bacteria) + Viruses**
    **3) all_reads: Eukaryota + Bacteria + Viruses + unclassified**
    
### Domain Classification 
```{r}
absolute_path = paste(path,"/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  domain_absolute <- current_sample %>% subset(V4 == "D" | V4 == "U") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))

  colnames(domain_absolute) <- c("Domain", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    domain_all_samples <-  domain_absolute

  } else {
    domain_all_samples <- full_join(domain_all_samples, domain_absolute,by="Domain")

  }
  i = i + 1
}

# save(domain_all_samples, file ="standard_absolute_domain.Rdata")

domain_all_samples[is.na(domain_all_samples)] <- 0
keys_to_gather <- colnames(domain_all_samples)[-1] # gather samples, exclude "Domain" from colnames

# prepare dataframe for ggplot
domain_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

order <- unique(domain_all_samples$samples)
domain_all_samples$samples <- factor(domain_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
domain_all_samples$Domain <- factor(domain_all_samples$Domain, level=unique(domain_all_samples$Domain))
domain_all_samples

```

```{r}


plot <- ggplot(domain_all_samples, aes(x=samples,y=Reads,fill=Domain)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Reads Classification with Standard Kraken2 DB (Domain)", y="Number of Reads", x = "Samples")+
  theme(axis.text.x=element_text(angle=90)) +
  scale_fill_manual(values= manual_color)
plot
# ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Absolute/kraken2_std_domain.png")

#### plot relative abundance 
domain_all_samples %>%
  group_by(samples) %>%
  mutate(Total=sum(Reads)) %>%
  mutate(Relative=Reads/Total) -> domain_relative


relative_plot <- ggplot(domain_relative, aes(x=samples,y=Relative,fill=Domain)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Relative Abundance with Standard Kraken2 DB (Domain)", y="Percentage", x = "Samples")+
  scale_y_continuous(labels = scales::percent_format()) +
  theme(axis.text.x=element_text(angle=90))+
  scale_fill_manual(values=manual_color)
relative_plot

# ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Relative/kraken2_std_domain_rel.png")
```

## From the results above, we can conclude that most of the reads in a metagenomic sample was not able to identified or belongs to the Eukaryota. 
    ***- The standard database only included human genome as Eukaryota but not the genomes of the real host, Rattus ***
    ***- thus many of the unclassified reads might belongs to nonoverlapping sequences between Rattus and human***


### Classify phylum, excluding unclassified and Eukaryote reads to increase local resolution
```{r}
absolute_path = paste(path,"/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  phylum_absolute <- current_sample %>% subset(V4 == "P") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))
  phylum_absolute <- phylum_absolute %>% subset(c(trimed!="Chordata")) %>%  subset(V2 >= 100)# deselect phylum belongs to Eukrayota
  colnames(phylum_absolute) <- c("Phylum", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    phylum_all_samples <-  phylum_absolute

  } else {
    phylum_all_samples <- full_join(phylum_all_samples, phylum_absolute,by="Phylum")

  }
  i = i + 1
}

# save(phylum_all_samples, file ="standard_absolute_phylum_micro_only.Rdata")

phylum_all_samples[is.na(phylum_all_samples)] <- 0
keys_to_gather <- colnames(phylum_all_samples)[-1] # gather samples, exclude "Phylum" from colnames

# prepare dataframe for ggplot
phylum_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

all_phylums <- unique(phylum_all_samples$Phylum) # all phylums classified
# phylums I want to factored into order so they can be assigned with a specific color during plotting
order_phylum <- c("Proteobacteria","Actinobacteria","Cyanobacteria","Firmicutes","Bacteroidetes","Uroviricota","Tenericutes","Spirochaetes","Fusobacteria","Chlamydiae","Aquificae","Chloroflexi","Thermotogae","Artverviricota","Peploviricota") 
remaining_phylum <- all_phylums %>% data.frame() %>% subset(!c(all_phylums %in% order_phylum)) 
remaining_phylum <- remaining_phylum$.

phylum_all_samples$samples <- factor(phylum_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
phylum_all_samples$Phylum <- factor(phylum_all_samples$Phylum, level=c(order_phylum,remaining_phylum))
phylum_all_samples %>% group_by(samples) %>%
  mutate(Total=sum(Reads), Percentage = Reads/Total) %>% subset(Percentage >= 0.001) ->major_phylum
```



```{r}



plot <- ggplot(major_phylum, aes(x=samples,y=Reads,fill=Phylum)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Reads Classification with Standard Kraken2 DB (Phylum)", y="Number of Reads", x = "Samples", caption = "* For the purpose of visualization, only taxa classified with more 1% in composition will be shown") + 
  theme(axis.title = element_text(size=1), axis.text.x = element_text(angle=90))+
  scale_fill_manual(values=manual_color)

plot

ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Absolute/kraken2_std_phylum.png", width = 8, height = 6)




relative_plot <- ggplot(major_phylum, aes(x=samples,y=Percentage,fill=Phylum)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Relative Abundance with Standard Kraken2 DB (Phylum)", y="Percentage", x = "Samples", caption = "* For the purpose of visualization, only taxa classified with more than 1% in composition will be shown") + 
  theme(axis.title = element_text(size=1), axis.text.x = element_text(angle=90), legend.key.size =unit(0.5,"cm"),plot.caption = element_text(hjust = 0, face= "italic")) +
  scale_y_continuous(labels = scales::percent_format()) +
  theme(axis.text.x=element_text(angle=90)) + 
  scale_fill_manual(values=manual_color)
relative_plot
# ggplotly(relative_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Relative/kraken2_std_phylum_rel.png",width = 8, height = 6)

```


### Classify genus, excluding unclassified and Eukaryote reads to increase local resolution
```{r}
absolute_path = paste(path,"/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  genus_absolute <- current_sample %>% subset(V4 == "G") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))
  genus_absolute <- genus_absolute %>% subset(c(trimed!="Homo")) %>% subset(V2 >= 100) # deselect genus belongs to Eukrayota
  colnames(genus_absolute) <- c("genus", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    genus_all_samples <-  genus_absolute

  } else {
    genus_all_samples <- full_join(genus_all_samples, genus_absolute,by="genus")

  }
  i = i + 1
}

# save(genus_all_samples, file ="standard_absolute_genus_micro_only.Rdata")

genus_all_samples[is.na(genus_all_samples)] <- 0
keys_to_gather <- colnames(genus_all_samples)[-1] # gather samples, exclude "genus" from colnames

# prepare dataframe for ggplot
genus_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

# genus I want to factored in order so they can be assigned with a specific color during plotting
all_genus <- unique(genus_all_samples$genus) # allgenus classified
order_genus <- c("Xanthomonas","Pseudomonas","Pasteurella","Streptomyces","Calothrix","Andhravirus","Mycoplasma","Spiroplasma","Bacillus","Staphylococcus","Lactobacillus","Clostridium","Mycobacterium","Bordetella","Leptospira","Yersinia","Escherichia","Bartonella","Bacteroides","Microcystis","Nostoc","Mycobacteroides","Alistipes","Parabacteroides","Anabaena","Dolichospermum","Faecalibacterium","Sphaerospermopsis","Prevotella","Achromobacter") 
remaining_genus <- all_genus %>% data.frame() %>% subset(!c(all_genus %in% order_genus)) # deselect genus subsetted to put in order
remaining_genus <- remaining_genus$. # data.frame to vector

order <- unique(genus_all_samples$samples)
genus_all_samples$samples <- factor(genus_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
genus_all_samples$genus <- factor(genus_all_samples$genus, level=c(order_genus,remaining_genus))
genus_all_samples %>% group_by(samples) %>%
  mutate(Total=sum(Reads), Percentage = Reads/Total) %>% subset(Percentage >= 0.010) ->major_genus
```

```{r}


plot <- ggplot(major_genus, aes(x=samples,y=Reads,fill=genus)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Reads Classification with Standard Kraken2 DB (Genus)", y="Number of Reads", x = "Samples", caption = "* For the purpose of visualization, only taxa classified with more than 1% in composition will be shown") + 
  theme(axis.title = element_text(size=1), axis.text.x = element_text(angle=90), legend.key.size=unit(0.5,"cm"),plot.caption = element_text(hjust = 0, face= "italic"),legend.box = unit(8,"cm")) +
  scale_fill_manual(values=manual_color)

plot
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Absolute/kraken2_std_genus.png",width = 10, height = 6)



relative_plot <- ggplot(major_genus, aes(x=samples,y=Percentage,fill=genus)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Relative Abundance with Standard Kraken2 DB (Genus)", y="Percentage", x = "Samples", caption = "*For the purpose of visualization, only taxa classified with more than 10% in composition will be shown") + 
  theme(axis.title = element_text(size=1), axis.text.x = element_text(angle=90), legend.key.size=unit(0.5,"cm"),plot.caption = element_text(hjust = 0, face= "italic"),legend.box = unit(8,"cm")) +
  scale_y_continuous(labels = scales::percent_format()) +
  theme(axis.text.x=element_text(angle=90)) + 
  scale_fill_manual(values=manual_color)
relative_plot
# ggplotly(relative_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Relative/kraken2_std_genus_rel.png", width = 10, height = 6)

```

## Phyloseq Package for Downstream Analysis

### Data Wrangling to format for phyloseq analysis
```{r}
absolute_path = paste(path,"/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  bacteria_profile <- current_sample %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V4,V2)) %>% subset(V4 == "D"| V4 == "K"|V4 == "P"|V4 == "C"|V4 == "O"| V4 == "F"| V4 == "G")# only keep rank that has certain taxonomy
  rownames(bacteria_profile) <- NULL # reorder rownames after subsetting
  # select all taxa below Bacteria, thus excluding unclassified, root and Eukaryota
  bacteria_profile <- bacteria_profile[as.numeric(rownames(bacteria_profile)) >= which(bacteria_profile["trimed"] == "Bacteria"),] 
  colnames(bacteria_profile) <- c("OTU","rank",unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    all_samples_w_taxa <-  bacteria_profile

  } else {
    all_samples_w_taxa <- full_join(all_samples_w_taxa, bacteria_profile,by=c("rank","OTU"))

  }
  i = i + 1
}

all_samples_w_taxa[is.na(all_samples_w_taxa)] <- 0 # all otu with NA reads fill in with 0
# some OTU has the same name in different rank, thus cause duplications
duplicated_otu <- all_samples_w_taxa$OTU[duplicated(all_samples_w_taxa$OTU)]

# make OTU with same name in different rank with rank indicator by adding "_rank"
all_samples_w_taxa$OTU <- ifelse(all_samples_w_taxa$OTU %in% duplicated_otu, paste(all_samples_w_taxa$OTU,all_samples_w_taxa$rank,sep="_"),all_samples_w_taxa$OTU)

# make the OTU names the rownames
rownames(all_samples_w_taxa) <- all_samples_w_taxa$OTU 
all_samples_w_taxa <- all_samples_w_taxa[,-1] 
```

```{r}
otu_tab <- all_samples_w_taxa[,-1] # get rid of the rank column to form an otu table
OTU <- otu_table(all_samples_w_taxa,taxa_are_rows = T)

```